### ISSUE TITLE
Simple tagging scheme

### ISSUE THREAD

**Issue by stevenbird** [aikuma/aikuma#455]
Implement a simple tagging scheme in which the tag string is stored as part of a filename in the user's `social` directory, and all tags are public.

There are four kinds of tag:
* language: a screen like the default language settings screen (with the green check button), e.g. `language-iso639_kor` or `language-mycustomlanguage`
* speaker: a speaker list screen, allows you to add a new speaker, e.g. `speaker-JSLFOEPSOVKW`
* discourse type: a panel of the 10 OLAC discourse type icons, e.g. `olac-discourse_narrative`
* custom: a text input box, e.g. `custom-i_like_this`

User interface:
* four more icons on the long-press drop-down for a file

* https://github.com/aikuma/aikuma/wiki/File-Storage
* http://www.language-archives.org/REC/discourse.html
* cf #67, #442

**Comment by stevenbird** [comment]
We have agreed to group these under a single menu option, for tagging, using this icon or similar: http://upload.wikimedia.org/wikipedia/commons/0/0c/VisualEditor_-_Icon_-_Tag.svg

**Comment by stevenbird** [comment]
Note that tags cannot contain whitespace.

**Comment by stevenbird** [comment]
Tags must be alphanumeric + underscore, and have a limit of 20 characters.

When a tag file is shared with the central Google drive, we need to call the update method, to associate this tag with the corresponding media file.

**Comment by stevenbird** [comment]
@hleeldc please add details

**Comment by hleeldc** [comment]
The following code is an example of adding tags to a recording. See #462 for more details on GoogleDriveIndex2.

```java
Index idx = new GoogleDriveIndex2(rootName, tokenManager, baseUrl, idToken);
Map<String,String> meta = new HashMap();
meta.put("tag1", "");    /* creates a tag "tag1" */
meta.put("k", "v");      /* creates a tag "k__v" */
idx.update(identifier, meta);
```

This demonstrates 2 ways of making tags. One way is to put tags as keys of the map ignoring the values. If non-empty value exists, a new tag is created by joining the key and the value with **two underscores**.

Tagged files can be searched as follows.

```java
idx.search("fullText contains 'k__v'", processor);
```

`processor` is an Index.SearchResultProcessor object which provides the `process` method. This method is passed a Map<String,String> object. In the above example, the map will contain pairs like the following. Note that the identifier is also returned.

- ("tag1", "")
- ("k", "v")
- ("identifier", "v01/testrnrmf@gmail_com/items/ualwfcixlzpu/ualwfcixlzpu-testrnrmf@gmail.com-source.wav")

---

As Steven said in the previous comment, tags should consist of alphanumeric characters and underscores. Other characters and punctuations will function as a delimiter. For example, if you tag a file with "orange-carrot", it has same effect as having two separate tags: "orange" and "carrot".

Keep tag length as short as possible and the tags storage is limited. Probably under 20 characters. Steven was able to store up to 700 characters, but I was able to store a little more that 1000 characters.

**Comment by stevenbird** [comment]
It might be dangerous to use double underscore, when underscore is a legal tag character.

**Comment by stevenbird** [comment]
@hleeldc at what time are this method called? When a file is synchronised with the private Google drive, or when a file is shared with the central Google drive? I'm guessing it is the latter.

Would you please confirm that the following is correct? (from #463):
1. user creates a file and tags it (creates a tag file in the phone storage and private drive)
2. user publicly shares a file (now we call this method to add the tag to the file description field)
3. user (maybe a different user) adds a tag to a shared file (creates a tag file in the phone storage)
4. user goes online

Now Aikuma calls the share method on the derivative file or tag file.

**Comment by hleeldc** [comment]
@stevenbird It can be called at either times. My example above assumes tagging while the file is already in the central google drive -- the example uses GoogleDriveIndex2 class, which sends the update request to the server who owns the central google drive.

But, it's possible to call the update method against the private google drive. In that case, GoogleDriveIndex should be used. This class is used when the aikuma folder is owned by the user. Note that the constructor parameter list is different as it doesn't have to rely on the remote server for updating.

```java
Index idx = new GoogleDriveIndex(rootName, tokenManager);
```

**Comment by hleeldc** [comment]
@stevenbird The double underscore is used in order to be able to parse out key and value from the tag. For example, one of the field in the metadata.json file is file_type. If single underscore was used, we would have created a tag like `file_type_source` by the client code calling the update method. Then, the search method would return a map entry `("file_type_source", "")` instead of `("file_type", "source")`.

**Comment by stevenbird** [comment]
@hleeldc – thanks for the explanation about double underscore.

The UI must restrict users to alphanumeric tags (possibly including space). Multiple spaces need to be trimmed (this will avoid accidentally creating double underscore).

**Comment by stevenbird** [comment]
@lisaslyis, please check that the following update method cannot cause `tag` to be duplicated in the file properties.

```
Index idx = new GoogleDriveIndex2(rootName, tokenManager, baseUrl, idToken);
Map<String,String> meta = new HashMap();
meta.put("tag", "");
idx.update(identifier, meta);
```

**Comment by stevenbird** [comment]
NB all tags have a key-value structure. Keys are: language, speaker, discourse-type, and custom. They will be stored as a compound string: `key__value` (double underscore) in the key position, with an empty value.

**Comment by stevenbird** [comment]
Resolved by #542
